### PR TITLE
table of application types and valid grant types

### PR DESCRIPTION
**Pull request by mystiberry-okta**
Resolves: OKTA-119934
test: bacon

@cwu-okta , @yuliu-okta , @johngronberg-okta Here is the table of values and requirements requested in OKTA-119934. Please review for completeness and correctness.

Note that a similar table for grant types and valid response types is being developed in another JIRA.

### CODE REVIEW

File: _source/_docs/api/resources/oauth-clients.md [hunk 1]
```diff
@@ -102,25 +102,38 @@ Client applications have the following properties:
...
-> All redirect URIs must be absolute URIs and must not include a fragment compontent.
+* All redirect URIs must be absolute URIs and must not include a fragment compontent.
+
+* Different application types have different valid values for the corresponding grant type:
+
+    |-------------------+---------------------------------------------------------------+-----------------------------------------------------------------------------------|
+    | Application Type  | Valid Grant Type                                              | Requirements                                                                      |
+    | ----------------- | ------------------------------------------------------------- | --------------------------------------------------------------------------------- |
+    | `web`             | `authorization_code`, implicit`, `refresh_token`              | Must have at least `authorization_code`                                           |
+    | `native`          | `authorization_code`, `implicit`, `password`, `refresh_token` | Must have either `authorization_code` or both `implicit` and `password`           |
```
Comment (yuliu-okta) [inline _source/_docs/api/resources/oauth-clients.md]: Must have at least `authorization_code`

Comment (mystiberry-okta) [inline _source/_docs/api/resources/oauth-clients.md]: @yuliu-okta The JIRA comment from @gronberg says either authorization_code OR implicit+password. Can you confirm it's just authorization_code that's required?

Comment (johngronberg-okta) [inline _source/_docs/api/resources/oauth-clients.md]: @yuliu-okta the following JSON will create a native client:

  ```
{
    "client_name": "Native test",
    "client_uri": null,
    "logo_uri": null,
    "redirect_uris": [
      "https://httpbin.org/get"
    ],
    "response_types": [
      "id_token",
      "token"
    ],
    "grant_types": [
      "password",
      "implicit"
    ],
    "token_endpoint_auth_method": "client_secret_basic",
    "application_type": "native"
  }
```

Comment (yuliu-okta) [inline _source/_docs/api/resources/oauth-clients.md]: @johngronberg-okta it looks like there is a gap between public client validator and the UI, based on the discussion [here](https://oktawiki.atlassian.net/wiki/display/UX/OAuth+Grant+types), we might need to change the validator.
BTW, in current implementation, you can even create a native client with only password(redirect_uris should be removed) or implicit(the token_endpoint_auth_method should be none).

Comment (johngronberg-okta) [inline _source/_docs/api/resources/oauth-clients.md]: ok, let's document the intended requirements, and then fix the validator bug.

Comment (mystiberry-okta) [inline _source/_docs/api/resources/oauth-clients.md]: So delete "or both `implicit` and `password` from requirement for `native` (line 127)?

Comment (yuliu-okta) [inline _source/_docs/api/resources/oauth-clients.md]: Yes, so it should be "Must have at least authorization_code"

**Comment by mystiberry-okta** [comment]
@yuliu-okta I think we've got it now!

**Comment by mystiberry-okta** [comment]
@yuliu-okta or @johngronberg-okta I need a "shipit"

**Comment by johngronberg-okta** [comment]
lgtm

**Pull request closed without merging**